phpBB/phpbb/auth/provider/db.php

```diff
@@ -153,8 +153,12 @@ public function login($username, $password)
 		}
 
 		$login_error_attempts = 'LOGIN_ERROR_ATTEMPTS';
-		$show_captcha = ($this->config['max_login_attempts'] && $row['user_login_attempts'] >= $this->config['max_login_attempts']) ||
-			($this->config['ip_login_limit_max'] && $attempts >= $this->config['ip_login_limit_max']);
+
+		$user_login_attempts	= $row ? ($this->config['max_login_attempts'] && $row['user_login_attempts'] >= $this->config['max_login_attempts']) : false;
```

Comment:
`$user_login_attempts = (is_array($row) && $this->config['max_login_attempts'] && $row['user_login_attempts'] >= $this->config['max_login_attempts']);`

Reply:
I think the way I proposed it is faster? Also used some lines above this changes.

`'user_id'				=> ($row) ? (int) $row['user_id'] : 0,`

Reply:
It shouldn't be faster, that would make zero sense, PHP needs to know the type either way, so type juggling cannot possibly be faster, but we are talking a couple of CPU cycles anyway, so it doesn't matter. `is_array` is explicit as to what you want to check.

Reply:
> We do not want to find whether the given variable is an array but a TRUE bool since a FALSE bool is already returned by the SQL query in case of false username.

Reply:
You have `$row['user_login_attempts']` in the code. This shouldn't work if `$row === true`. What am I missing?

Reply:
It is a ternary operatior `$row ? etc : false;` if **$row** is TRUE then evaluate the logic else return **false*

Reply:
This query
```php
		$sql = 'SELECT *
			FROM ' . USERS_TABLE . "
			WHERE username_clean = '" . $this->db->sql_escape($username_clean) . "'";
		$result = $this->db->sql_query($sql);
		$row = $this->db->sql_fetchrow($result);
		$this->db->sql_freeresult($result);
```

Returns falkse in case, as I said. But returns everything we need otherwise, as an array already.

Reply:
![image](https://user-images.githubusercontent.com/480857/112221333-97017380-8c27-11eb-91af-44e9c1adc319.png)

Reply:
I don't really want to keep arguing, so here are my thoughts:
* the conditional operator cannot possibly be faster, as you first implicitly cast a variable to `bool`, which must require type look up, otherwise it cannot work, and thus `is_array` cannot possibly have any overhead;
* `is_array` is the most explicit way of checking that `$row` is an array, which is the thing you want to check, so you can index into it;
* your change is unnecessarily verbose, if it was `(1 == 1) ? true : false;` we surely could agree, that `(1 == 1)` is good enough;
* In this case, there is no need to branch, as you have nothing in the `else` branch, but `false`, hence, you want to compute a boolean expression;
* There is a shorter way to write the expression above, that is what I gave you;
* using `&&` is enough, the rest of the expression won't be evaluated - see `var_dump(false && 2 / 0);`;
* the way any code is written in this file or project is a poor argument to justify your changes, as you are changing the code. Surely, if all the code was awesome, you wouldn't need to make any changes

Reply:
Thanks for your time.